### PR TITLE
Release process

### PR DESCRIPTION
**Pull request by lehins**
# Description

What I've learned while being on duty as a release engineer.

This is just an initial version and is meant to be a live document.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the `CHANGELOG.md` for affected package
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: RELEASING.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+## Release Process
+
+Usually a package release will happen directly from `master` branch. Current release
+process even allows for a release to happen from a commit that is slightly behind `HEAD`
+of `master` branch, but every major release must happen from `master`. On a rare occasion
+when a bugfixes need to be backported to an older version, a release can happen from an
+ephemeral release branch. Such a release branch must follow the same procedure of a PR
+review and CI, that is why it should be prefixed with `release/some-branch-name`, cause
+taht will trigger the CI. should be followed during a bugfix release to an older version,
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
that will trigger the CI. should be followed during a bugfix release to an older version,
```
Also seems like there's something missing from the beginning of the sentence.

File: RELEASING.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+that has diverged from master, as for the PR into master directly.
+
+### Release Steps
+
+Here are the steps for release engineer to folow.
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
Here are the steps for release engineer to follow.
```

File: RELEASING.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+#### Decide what to release
+
+Normally it should be OK to release at the same time all packages from the repository that
+experienced some sort of change, but that is not mandatory. Any package can be released by
+itself at any point from `HEAD` of `master` or a even commit that is behind the `HEAD` of
+`master`. A release with minor version bump from an ephemeral release branch is also
+possible.
+
+The most common case is to blindly release all of the packages that were changed since
+they werre last released. This begs a question. How to decide which of the packaged have
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
they were last released. This begs a question. How to decide which of the packages have
```

File: RELEASING.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+changed, thus deserve a release?
+
+The rule is very simple. Every package that falls under this release process and has a
+version in its cabal file higher than the highest version released to CHaPs is allowed to
+be released. It sis possibly to rely on git tags for information about the latest released
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
be released. It is possible to rely on git tags for information about the latest released
```

File: RELEASING.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+   used as base:
+
+   ```shell
+   $ git checkout -b release/cardano-ledger-core-1.21.3.0 cardano-ledger-core-1.21.2.1
+   ```
+
+2. Changes that need to be released should be `cherry-pick`ed from master. If a fix on
+   `master` was implemented in some incompatible fashion to the current release, then it
+   is fine to reimplement it anew, as long as the change being introduced is also present
+   on master in some form. That also includes a changelog entry, it should be presetn in
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
   on master in some form. That also includes a changelog entry, it should be present in
```

File: RELEASING.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+
+## Packages excluded from release process
+
+We release most of the packages in this repo to [CHaP (Cardano Haskell
+Packages)](https://github.com/input-output-hk/cardano-haskell-packages). However, there
+are a few packages that are either used for testing, debugging or benchmarking and do not
+deserve to be released into the World. They are neither released nor versioned. Bounds on
+the local depndencies do not need to be updated, because they will always use the versions
+for ledger packages from whith the repositiory, rather than from CHaPs. Here is the full
```
Comment (Soupstraw) [inline RELEASING.md]:
```suggestion
for ledger packages from within the repositiory, rather than from CHaPs. Here is the full
```

**Review by Soupstraw — COMMENTED** [review]
Great job, that's some very helpful documentation. I think it'd be a good idea to link to the RELEASING file from CONTRIBUTING (or even better - write a summary of the information that's relevant to every contributor), since everyone has to think about version bumps, not just the release engineer.

File: RELEASING.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+```
+# Version history for `cardano-ledger-core`
+
+## 1.20.2.0
+
+* Add ...
+...
+```
+
+It is quite common to experience conflicts in the changelog, since that will be the most
```
Comment (michaelpj) [inline RELEASING.md:73]: You might enjoy using `scriv`, we've been using it fairly happily in `plutus` to solve this.

Comment (lehins) [reply RELEASING.md:73]: Nah, `scriv` is really a hammer that we don't need. It would be very useful if we were dealing with bundles as Plutus or Consensus does, but we don't have this problem. Moreover, we as developers, deal with git conflicts on daily basis, so couple more conflicts in the changelog is not a big deal.

File: RELEASING.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+upper bound on `containers` package, because that's just dumb. On the other hand all of
+the dependeices that come from CHaPs should follow strict bounds, because those packages
+are a lot more volatile and creating revisions for packages on CHaPs is not as smooth as
+it is on Hackage.
+
+### `CHANGELOG.md`
+
+Every package, that [is subject to the release
+process](packages-excluded-from-release-process), will always have the top most entry in
+the `CHANGELOG.md` set to an unreleased version. That top most section will have either
```
Comment (JaredCorduan) [inline RELEASING.md:56]: Did we agree to putting "(unreleased)" next to the top entry?

For example:

```
1.2.0 (unreleased)
* stuff

1.1.0
* things
```

This way, when `1.2.0` is released, we do not need to do something like add a new `1.2.1` or some such as a part of release process (to maintain the invariant that we "always have the top most entry in the `CHANGELOG.md` set to an unreleased version").

File: RELEASING.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+  ### `testlib`
+
+  * Add `Arbitrary` instance for ...
+  ...
+  ```
+
+## Release Process
+
+Usually a package release will happen directly from `master` branch. Current release
```
Comment (JaredCorduan) [inline RELEASING.md]: We should remove mention of the "trunk based developement" then: https://github.com/input-output-hk/cardano-ledger/blob/master/CONTRIBUTING.md#branching-model

File: RELEASING.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+4. Once the package has been released and a git tag for that release was created, the
+   `release/` branch can be removed.
+
+This process does not accomodate backporting fixes to versions that are at least two major
+versions behind the one on `master`.
+
+### Deciding what to realease
+
+Figure out what's dirty is hard and should be improved
```
Comment (JaredCorduan) [inline RELEASING.md]:
```suggestion
Figure out what's dirty is hard and should be improved.
```

File: RELEASING.md (hunk 8 above)
Comment (JaredCorduan) [reply RELEASING.md:56]: chatted in a call, I like the new plan!

File: RELEASING.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+With respect to the actual change log entries, there is no enforced order in which entries
+should be added. To the top, to the bottom or in the middle of the list, it is
+irrelevant. In fact, randomized insertion reduces conflicts, so feel free to add entries
+anywhere in the section. Most important is that we can always infer the actual order from
+the git history, which provide with a very reliable source of order on changes.
+
+#### When to update
+
+`CHANGELOG.md` must be updated for the package that is expereincing a change and any other
```
Comment (aniketd) [inline RELEASING.md]:
```suggestion
`CHANGELOG.md` must be updated for the package that is experiencing a change and any other
```

File: RELEASING.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+3. Regular release process should follow from here.
+
+4. Once the package has been released and a git tag for that release was created, the
+   `release/` branch can be removed.
+
+This process does not accomodate backporting fixes to versions that are at least two major
+versions behind the one on `master`.
+
+### Deciding what to realease
```
Comment (aniketd) [inline RELEASING.md]:
```suggestion
### Deciding what to release
```

File: RELEASING.md (hunk 1 above)
Comment (lehins) [reply RELEASING.md]: Fixed. Thanks

File: RELEASING.md (hunk 9 above)
Comment (lehins) [reply RELEASING.md]: I've removed the link and other no longer relevant information. I have left a mention about the trunk based development, since technically that is still our model. We just do releasing differently, since we have many packages that we can release, not just one.

**Comment by lehins** [comment]
> I think it'd be a good idea to link to the RELEASING file from CONTRIBUTING

Great idea. Made a small section with a link. There is also some info about versioning info in the [Cardano Engineering Handbook](https://github.com/input-output-hk/cardano-engineering-handbook/blob/main/policy/haskell/packaging/versioning.md), which I linked to in the RELEASEING doc as well.

> since everyone has to think about version bumps, not just the release engineer.

That is true, but I think they are very much related and it is because of the release process all of those changes to versions and bounds are necessary. Moreover, I believe, everyone on the team should know how the release process works, so it is relevant to everyone

File: RELEASING.md [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+a lower and, potentially, an upper bound update on the dependency that has experienced the
+change.
+
+For example if I add a function to `cardano-ledger-core` and use it in
+`cardano-ledger-conway`, then I must add a minor version bump to `cardano-ledger-core` and
+I also must add a lower version restriction in `cardano-ledger-conway` for
+`cardano-ledger-core`, as per PVP.
+
+In other words, when it comes to versioning, the fact that all of the packages in this
+repository are build together at the same time should be ignored. It should be assumed
```
Comment (teodanciu) [inline RELEASING.md]: Just a little nitpick:  `are built together`

File: RELEASING.md [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+bounds for dependencies that come from Hackage. For example, there is no need to add an
+upper bound on `containers` package, unless there is a known incompatibility. Speculative
+upper bounds over time lead to overly restrictive build plans that are forced to favor
+older packages. Moreover, they lead to a lot of redundant maintenance burden. That being
+said, most of the packages from within the same repository are a lot more volatile and
+require tight interoperability, therefore speculative upper bounds are encouraged.
+
+### Lower bounds
+
+Placing lower bounds is a lot less controversial than upper bounds, because the don't
```
Comment (teodanciu) [inline RELEASING.md]: `they` don't suffer

File: RELEASING.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+suffer from speculation about the future. Older versions of dependencies are known at
+release time. Therefore lower bounds must be added for all dependencies that are known to
+be incompatible with the package
```
Comment (teodanciu) [inline RELEASING.md:45]: So, say we make  change in `cardano-core` , which `cardano-ledger-babbage` depends on, but this change is not directly impacting `cardano-ledger-babbage` ( so nothing has to be changed in the `cardano-ledger-babbage` to acccomodate the change). 
In this case we leave the lower bound of `cardano-ledger-babbage` as it is, correct?

**Review by teodanciu — APPROVED** [review]
Fantastic explanations!  Thanks for making such a daunting topic so clear.

File: RELEASING.md (hunk 15 above)
Comment (lehins) [reply RELEASING.md:45]: That is absolutely correct.

File: RELEASING.md (hunk 13 above)
Comment (lehins) [reply RELEASING.md]: That's not a nitpick but a legit spelling mistake. :slightly_smiling_face: Thank you.

**Review by aniketd — APPROVED** [review]
Awesome write up!

File: RELEASING.md (hunk 9 above)
Comment (JaredCorduan) [reply RELEASING.md]: ok, thanks!

**Review by JaredCorduan — APPROVED** [review]
this is awesome, thank you so much!

File: RELEASING.md [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+Normally, [upper bounds are required by
+PVP](https://pvp.haskell.org/#dependencies-in-cabal) for all dependencies of a
+package. This is the only exception we make to the PVP, and we do not require strict upper
+bounds for dependencies that come from Hackage. For example, there is no need to add an
+upper bound on `containers` package, unless there is a known incompatibility. Speculative
+upper bounds over time lead to overly restrictive build plans that are forced to favor
+older packages. Moreover, they lead to a lot of redundant maintenance burden. That being
+said, most of the packages from within the same repository are a lot more volatile and
+require tight interoperability, therefore speculative upper bounds are encouraged.
```
Comment (TimSheard) [inline RELEASING.md]: We are taking about the ledger-state repository when it says "within the same repository" right?
So might it be better to say,  we encourage upper bounds on packages within the ledger-state repository.

**Review by TimSheard — APPROVED** [review]
A lot of work went into this. Thank for doing it.

File: RELEASING.md (hunk 16 above)
Comment (lehins) [reply RELEASING.md]: Yeah, you are right, I'll reword it to be more ledger specific.

Comment (lehins) [reply RELEASING.md]: Done
